### ISSUE TITLE
PyTorch adversarial attack strange behavior 

### ISSUE THREAD

**Issue by metaluga145** [IBM/adversarial-robustness-toolbox#89]
**Describe the bug**
Modified example produces strange output. Please see code and images below.

**To Reproduce**
1. Run the code below
2. Save produced adversarial images from test set
3. See produced outputs 

```python
"""Trains a convolutional neural network on the MNIST dataset, then attacks it with the FGSM attack."""
from __future__ import absolute_import, division, print_function, unicode_literals

import sys
from os.path import abspath
import cv2

sys.path.append(abspath('.'))

import keras.backend as k
from keras.models import Sequential
from keras.layers import Dense, Flatten, Conv2D, MaxPooling2D, Dropout
import torch
from torch import nn


import numpy as np

from art.attacks.fast_gradient import FastGradientMethod
from art.classifiers import KerasClassifier, PyTorchClassifier
from art.utils import load_dataset

# Read MNIST dataset
(x_train, y_train), (x_test, y_test), min_, max_ = load_dataset(str('mnist'))

class Flatten(nn.Module):
    def forward(self, x):
        return x.view(x.size()[0], -1)

class MNIST_net(nn.Module):
    def __init__(self):
        super().__init__()
        self.features = nn.Sequential(
                            nn.Conv2d(1, 32, kernel_size=(3, 3), padding=(1, 1)),
                            nn.ReLU(),
                            nn.Conv2d(32, 32, kernel_size=(3, 3), padding=(1, 1)),
                            nn.ReLU(),
                            nn.MaxPool2d(kernel_size=2, stride=2),
                            nn.Dropout(0.25),
                            Flatten()
                        )
        self.classifier = nn.Sequential(
                            nn.Linear(6272, 128),
                            nn.ReLU(),
                            nn.Dropout(0.5),
                            nn.Linear(128, 10)
                          )

    def forward(self, x):
        out = self.features(x)
        logits = self.classifier(out)
        return logits


x_train = x_train.transpose(0, 3, 1, 2)
x_test = x_test.transpose(0, 3, 1, 2)

# Create Keras convolutional neural network - basic architecture from Keras examples
# Source here: https://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py
# k.set_learning_phase(1)
# model = Sequential()
# model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=x_train.shape[1:]))
# model.add(Conv2D(64, (3, 3), activation='relu'))
# model.add(MaxPooling2D(pool_size=(2, 2)))
# model.add(Dropout(0.25))
# model.add(Flatten())
# model.add(Dense(128, activation='relu'))
# model.add(Dropout(0.5))
# model.add(Dense(10, activation='softmax'))

# model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# classifier = KerasClassifier(clip_values=(min_, max_), model=model, preprocessing=(0.1307, 0.3081))
# classifier.fit(x_train, y_train, nb_epochs=5, batch_size=128)

model = MNIST_net()
classifier = PyTorchClassifier(
    model,
    torch.nn.CrossEntropyLoss(),
    torch.optim.Adam(model.parameters()),
    (1, 28, 28),
    10,
    preprocessing=(0.1307, 0.3081)
)

classifier.fit(x_train, y_train, nb_epochs=5, batch_size=64)

# Evaluate the classifier on the test set
preds = np.argmax(classifier.predict(x_test), axis=1)
acc = np.sum(preds == np.argmax(y_test, axis=1)) / y_test.shape[0]
print("\nTest accuracy: %.2f%%" % (acc * 100))
```

**Expected behavior**
The adversarial example is expected to have smaller perturbation magnitude and be similar to Keras version.

**Screenshots**
Original image(MNIST test image 999)
![original_9](https://user-images.githubusercontent.com/9748910/59122111-902a4380-8962-11e9-96de-6f6e992d9023.png)

PyTorch result
![PyTorch_result_9](https://user-images.githubusercontent.com/9748910/59121997-37f34180-8962-11e9-9dbe-b9703e2c0026.png)

Keras result
![Keras_result](https://user-images.githubusercontent.com/9748910/59122205-d8e1fc80-8962-11e9-9b68-19cddf380762.png)

**System information (please complete the following information):**
 - OS: Ubuntu 16.04
 - Python version: 3.7
 - ART version or commit number: from PyPi
 - PyTorch version: 1.0.0 and 0.4.0

**Comment by ririnicolae** [comment]
@metaluga145 Thanks for reporting this, we'll look into it! What is the test accuracy obtained by your model after training on clean samples?

**Comment by metaluga145** [comment]
@ririnicolae Test accuracy after training on clean samples is 97.43%.
Test accuracy on adversarial sample: 89.40%

**Comment by beat-buesser** [comment]
Hi @metaluga145 Sorry for the late response. I did a test run with the script above and noticed that the `PyTorchClassifier` does not have the argument `clip_values` set while the `KerasClassifier` does have it. Without `clip_values` the attack can generate adversarial images outside of the pixel range [min_, max_]. If I set `clip_values=(min_, max_)` in the `PyTorchClassifier` I get results comparable to the ones from `KerasClassifier` with the script above.
